File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        
        self.maxPath = float("-inf")
        def dfs(node):
            if not node:
                return 0

            left = dfs(node.left)
            right = dfs(node.right)
            nodePath = max(left + node.val, right + node.val, left + right + node.val, node.val)
            self.maxPath = max(nodePath, self.maxPath)
            return max(max(left, right) + node.val, node.val)

        dfs(root)
        return self.maxPath

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private:
    int cnt = 0;
public:
    int kthSmallest(TreeNode* root, int k) {
        if (!root){
            return -1;
        }
        int ret = -1;
        ret = kthSmallest(root->left, k);
        if (ret != -1){
            return ret;
        }
        if (++cnt == k){
            return root->val;
        }
        ret = kthSmallest(root->right, k);
        if (ret != -1){
            return ret;
        }
        return -1;
    }
};
